panic: dial udp 8.8.8.8:80: connect: network is unreachable

./manager-api   内网不能联外网，启动提示connect: network is unreachable
panic: dial udp 8.8.8.8:80: connect: network is unreachable
goroutine 1 [running]:
github.com/apisix/manager-api/internal/utils.GetOutboundIP(0x0, 0x0, 0x0)
	/root/apisix-dashboard/api/internal/utils/utils.go:61 +0x133
github.com/apisix/manager-api/internal/utils.init.0.func1(0x0, 0xecb95ae00, 0x0)
	/root/apisix-dashboard/api/internal/utils/utils.go:41 +0x2b
github.com/sony/sonyflake.NewSonyflake(0x0, 0x0, 0x0, 0xc0000e5eb0, 0x0, 0x3)
	/root/go/pkg/mod/github.com/sony/sonyflake@v1.0.0/sonyflake.go:75 +0x14a
github.com/apisix/manager-api/internal/utils.init.0()
	/root/apisix-dashboard/api/internal/utils/utils.go:39 +0xb4

--- a/api/internal/utils/utils.go
+++ b/api/internal/utils/utils.go
@@ -18,6 +18,7 @@ package utils
 
 import (
 	"fmt"
+	"math/rand"
 	"net"
 	"os"
 	"strconv"
@@ -37,10 +38,13 @@ func init() {
 		}
 		salt = uint16(i)
 	}
-
+	ips, err := getLocalIPs()
+	if err != nil {
+		panic(err)
+	}
 	_sf = sonyflake.NewSonyflake(sonyflake.Settings{
 		MachineID: func() (u uint16, e error) {
-			return sumIP(GetOutboundIP()) + salt, nil
+			return sumIP(randomIP(ips)) + salt, nil
 		},
 	})
 	if _sf == nil {
@@ -50,22 +54,34 @@ func init() {
 
 func sumIP(ip net.IP) uint16 {
 	total := 0
-	for i := range ip {
-		total += int(ip[i])
+	if ip != nil {
+		for i := range ip {
+			total += int(ip[i])
+		}
 	}
 	return uint16(total)
 }
 
-// Get preferred outbound ip of this machine
-func GetOutboundIP() net.IP {
-	conn, err := net.Dial("udp", "8.8.8.8:80")
+func getLocalIPs() ([]net.IP, error) {
+	var ips []net.IP
+	addrs, err := net.InterfaceAddrs()
 	if err != nil {
-		panic(err)
+		return ips, err
 	}
-	defer conn.Close()
+	for _, a := range addrs {
+		if ipNet, ok := a.(*net.IPNet); ok && !ipNet.IP.IsLoopback() && ipNet.IP.To4() != nil {
+			ips = append(ips, ipNet.IP)
+		}
+	}
+	return ips, nil
+}
 
-	localAddr := conn.LocalAddr().(*net.UDPAddr)
-	return localAddr.IP
+func randomIP(xs []net.IP) net.IP {
+	if xs != nil && len(xs) > 0 {
+		return xs[rand.Intn(len(xs))]
+	} else {
+		return nil
+	}
 }
 
 func GetFlakeUid() uint64 {

--- a/api/internal/utils/utils.go
+++ b/api/internal/utils/utils.go
@@ -44,7 +44,7 @@ func init() {
 	}
 	_sf = sonyflake.NewSonyflake(sonyflake.Settings{
 		MachineID: func() (u uint16, e error) {
-			return sumIP(randomIP(ips)) + salt, nil
+			return sumIPs(ips) + salt, nil
 		},
 	})
 	if _sf == nil {
@@ -52,9 +52,9 @@ func init() {
 	}
 }
 
-func sumIP(ip net.IP) uint16 {
+func sumIPs(ips []net.IP) uint16 {
 	total := 0
-	if ip != nil {
+	for _, ip := range ips {
 		for i := range ip {
 			total += int(ip[i])
 		}

--- a/api/internal/utils/utils.go
+++ b/api/internal/utils/utils.go
@@ -18,7 +18,6 @@ package utils
 
 import (
 	"fmt"
-	"math/rand"
 	"net"
 	"os"
 	"strconv"
@@ -74,14 +73,6 @@ func getLocalIPs() ([]net.IP, error) {
 		}
 	}
 	return ips, nil
-}
-
-func randomIP(xs []net.IP) net.IP {
-	if xs != nil && len(xs) > 0 {
-		return xs[rand.Intn(len(xs))]
-	} else {
-		return nil
-	}
 }
 
 func GetFlakeUid() uint64 {

--- a/api/internal/utils/utils_test.go
+++ b/api/internal/utils/utils_test.go
@@ -42,11 +42,3 @@ func TestSumIPs_with_nil(t *testing.T) {
 	total := sumIPs(nil)
 	assert.Equal(t, uint16(0), total)
 }
-
-func TestRandomIP(t *testing.T) {
-	ips, err := getLocalIPs()
-	assert.Equal(t, nil, err)
-	ip := randomIP(ips)
-	t.Log(ip)
-	assert.Contains(t, ips, ip)
-}

--- a/api/internal/utils/utils_test.go
+++ b/api/internal/utils/utils_test.go
@@ -38,8 +38,8 @@ func TestGetLocalIPs(t *testing.T) {
 	assert.Equal(t, nil, err)
 }
 
-func TestSumIP_with_nil(t *testing.T) {
-	total := sumIP(nil)
+func TestSumIPs_with_nil(t *testing.T) {
+	total := sumIPs(nil)
 	assert.Equal(t, uint16(0), total)
 }
 

--- a/api/internal/utils/utils_test.go
+++ b/api/internal/utils/utils_test.go
@@ -17,8 +17,9 @@
 package utils
 
 import (
-	"github.com/stretchr/testify/assert"
 	"testing"
+
+	"github.com/stretchr/testify/assert"
 )
 
 func TestGetFlakeUid(t *testing.T) {
@@ -29,4 +30,23 @@ func TestGetFlakeUid(t *testing.T) {
 func TestGetFlakeUidStr(t *testing.T) {
 	id := GetFlakeUidStr()
 	assert.NotEqual(t, "", id)
+	assert.Equal(t, 18, len(id))
+}
+
+func TestGetLocalIPs(t *testing.T) {
+	_, err := getLocalIPs()
+	assert.Equal(t, nil, err)
+}
+
+func TestSumIP_with_nil(t *testing.T) {
+	total := sumIP(nil)
+	assert.Equal(t, uint16(0), total)
+}
+
+func TestRandomIP(t *testing.T) {
+	ips, err := getLocalIPs()
+	assert.Equal(t, nil, err)
+	ip := randomIP(ips)
+	t.Log(ip)
+	assert.Contains(t, ips, ip)
 }

